C typesupport

dirk-thomas opened the pull request:
Currently trying to make Windows happy...

tfoote commented:
@wjwwood says that you're not exporting symbols; that's why you don't get a .LIB with your .DLL.

dirk-thomas commented:
The problem that it needs to be:
* in a single compilation unit in order to be able to compare the pointer instead of the string value
* visible in the the library in order to use the symbol in other code parts
* a compile-time constant in order to use it as in a static initializer list

I have not yet found a way to achieve all of this at the same time on Windows.

gerkey commented:
As we discovered earlier today, in C you just can't use static const variables in static initializers.  E.g., this code doesn't compile with gcc on Linux:

~~~
static const int a = 0;
static const int b = a;
~~~

...but it does compile with g++.

dirk-thomas commented:
This will be continued once ros2/rmw_implementation#9 has been completed.

dirk-thomas commented:
I have rebased and updated this - ready for review:
* http://ci.ros2.org/job/ci_linux/734/
* http://ci.ros2.org/job/ci_osx/610/
* http://ci.ros2.org/job/ci_windows_opensplice/803/
* http://ci.ros2.org/job/ci_windows_connext_static/64/
* http://ci.ros2.org/job/ci_windows_connext_dynamic/52/

Locally I got it to compile and link the following example: https://github.com/ros2/examples/compare/master...c_msgs but since it is a weird mixture of C and C++ it won't run.

I think this should be merged even if we don't have a working talker / listener example in C with rcl(c) yet.

wjwwood commented on rosidl_typesupport_introspection_c/include/rosidl_typesupport_introspection_c/message_introspection.h:
```diff
@@ -0,0 +1,55 @@
+#include "rosidl_typesupport_introspection_c/visibility_control.h"
+
+typedef struct
+// ROSIDL_TYPESUPPORT_INTROSPECTION_C_PUBLIC
```
Should this be dropped?

dirk-thomas replied:
Yeah, I removed them for now. We can put them in when we need them down the road.

wjwwood commented on rosidl_typesupport_introspection_c/include/rosidl_typesupport_introspection_c/message_introspection.h:
```diff
@@ -0,0 +1,55 @@
+} rosidl_typesupport_introspection_c__MessageMember;
+
+typedef struct
+// ROSIDL_TYPESUPPORT_INTROSPECTION_C_PUBLIC
```
Same here?

wjwwood commented:
lgtm
